bbq/compiler/compiler.go

```diff
@@ -765,6 +789,26 @@ func (c *Compiler[_, _]) reserveFunctionGlobals(
 			members.Functions(),
 			members.Composites(),
 			members.Interfaces(),
+			members.Attachments(),
+		)
+	}
+
+	for _, declaration := range attachmentDecls {
+		compositeType := c.DesugaredElaboration.CompositeDeclarationType(declaration)
+		// Reserve a global for the constructor function.
+
+		constructorName := commons.TypeQualifier(compositeType)
+		c.addGlobal(constructorName)
+
+		members := declaration.Members
+
+		c.reserveFunctionGlobals(
+			compositeType,
+			members.SpecialFunctions(),
+			members.Functions(),
+			members.Composites(),
+			members.Interfaces(),
+			members.Attachments(),
```

Comment:
Reserving of global variables must be done in the same order as they are visited at the top-level (in `Compile()` method ) as well as when visiting as members of other composites (`compileCompositeMembers`). Given in other two places (`Compile()` and `compileCompositeMembers()`) attachments are visited after interfaces, we need to visit them in the same order here. Otherwise index lookup can get messed-up.

Unfortunately I couldn't come up with a better way to enforce this visiting-order in a non-heuristic way. If you have any ideas please feel free to add/suggest 🙏